### PR TITLE
Operation state machine

### PR DESCRIPTION
**Pull request by bergundy**
## Note for reviewers

Because this PR includes unstable API changes, the base branch is `nexus`, we'll wait until we've tested the API before merging both api and server repo changes to main.

## What changed?

- Implemented the state machine definitions for Nexus Operations including cancelation.
- Added a `Concurrent()` method on the task interface to mark tasks as concurrent and have them opt out of transition count match for staleness checks.

I considered making another machine node just for the schedule-to-close task and not adding the concurrent abstraction.
I'm a bit on the fence which approach would be better. I can submit a commit that shows the alternative.

## Why?

People want to invoke operations from workflows, this is the first step for enabling that.

## How did you test it?

Added unit tests.

### CODE REVIEW

File: service/history/hsm/sm.go
Comment (yycptt) [inline service/history/hsm/sm.go:37]: this is for allowing the implementation to have access to child state machines?

**Review by yycptt — APPROVED** [review]
LGTM! But I may not know the expected operation state machine behavior good enough.

**Review by MichaelSnowden — APPROVED** [review]

File: service/history/hsm/sm.go
Comment (bergundy) [reply service/history/hsm/sm.go:37]: Or parents, probably more for parents but could apply to both.

**Comment by bergundy** [comment]
I'll wait for the review of #5546 before merging.

**Pull request merged**
